### PR TITLE
getOrPut extension function for maps (Discussion #341)

### PR DESCRIPTION
**Pull request by Quillraven**
So far I am only missing this function in `ObjectMap` but I added it now also to several other map types.

As always, open for feedback :)

### CODE REVIEW

**Comment by Quillraven** [comment]
no idea how to fix the conflict since a pull tells me I am up to date and I can push without a conflict ... In my opinion my version should be the correct one.

**Comment by czyzby** [comment]
It says that: 

> This branch is 5 commits ahead, 7 commits behind libktx:develop.

Are you sure you've pulled the latest changes from `develop` from the `libktx/ktx` repository?

File: collections/README.md
Comment (czyzby) [inline collections/README.md:99]: Describe the behavior briefly, you don't need to mention stdlib.

File: collections/src/main/kotlin/ktx/collections/maps.kt
Comment (czyzby) [inline collections/src/main/kotlin/ktx/collections/maps.kt:433]: This will not work correctly if LibGDX maps can store `null` values. It would be better to check if the key is present in the map. Please add a test case for this.

Comment (czyzby) [reply collections/src/main/kotlin/ktx/collections/maps.kt:433]: Same goes for other maps.

File: collections/src/test/kotlin/ktx/collections/MapsTest.kt
Comment (czyzby) [inline collections/src/test/kotlin/ktx/collections/MapsTest.kt:403]: Ensure that the value was not replaced in the map, i.e. `"42"` should still point to `42`.

Comment (czyzby) [reply collections/src/test/kotlin/ktx/collections/MapsTest.kt:403]: Same goes for other similar tests.

File: collections/src/test/kotlin/ktx/collections/MapsTest.kt
Comment (czyzby) [inline collections/src/test/kotlin/ktx/collections/MapsTest.kt:473]: Add test cases with `null` values.

**Review by czyzby — CHANGES_REQUESTED** [review]
Try to resolve the conflict in your repo - pull the latest changes from this repository, as it seems that a few commits are missing. Ensure the extensions work as expected given null values.

File: collections/src/main/kotlin/ktx/collections/maps.kt
Comment (Quillraven) [reply collections/src/main/kotlin/ktx/collections/maps.kt:433]: The reason I went for that approach is to avoid a second call to `locateKey` which should be slighty more efficient than this approach:

```Kotlin
if(key in map) {
  return map[key] // second time locateKey is called
} else {
  // ...
}
```

But, as always, you are correct that this will not work in case you put null values to the map. I did not consider that and will change it.

Comment (czyzby) [reply collections/src/main/kotlin/ktx/collections/maps.kt:433]: > The reason I went for that approach is to avoid a second call to l`ocateKey` which should be slighty more efficient

That's what I thought. You can (kind of) avoid the second call with something along the lines of:

```kotlin
val value = map[key]
if (value == null && key !in map) {
  // ...
}
```
In best case scenario, it will perform only a single lookup and return the existing value.

Comment (czyzby) [reply collections/src/main/kotlin/ktx/collections/maps.kt:433]: Now that I think about, what happens if you try to use a `null` key? It's OK if it throws an NPE error - you don't need to change the function itself, but maybe an extra test for the thrown exception would make sense just to document the behavior.

Comment (Quillraven) [reply collections/src/main/kotlin/ktx/collections/maps.kt]: How can I test that? :D Key is a not nullable type and therefore I cannot pass null?

Comment (czyzby) [reply collections/src/main/kotlin/ktx/collections/maps.kt]: Kotlin type system strikes again, we might not have to test for this after all.

Anyway, you could explicitly define the key type:

```kotlin
val map = gdxMapOf<String?, String?>()
map.put(null, null) // Compiles.
```

**Comment by Quillraven** [comment]
I think I updated the branch now correctly. I pulled from the "upstream" develop branch and got some updates now and also the conflict (which is now resolved).

Also, added now the null test cases in case a map stores null values.

File: collections/src/main/kotlin/ktx/collections/maps.kt
Comment (Quillraven) [reply collections/src/main/kotlin/ktx/collections/maps.kt]: That is true but does that ever make sense? Why would I use a nullable key in Kotlin?

Comment (czyzby) [reply collections/src/main/kotlin/ktx/collections/maps.kt]: The only reason I can think of is receiving a Java type (`Any!`) and falsely assuming is not null. You can easily avoid this is in pure Kotlin code, but interacting with Java can lead to weird errors like think.

Since type system mostly prevents you from making this error, I'll leave it up to you. If you think the extra tests could be helpful, include them in the PR. Otherwise you can resolve the discussion.

**Review by czyzby — APPROVED** [review]
We can merge this as is after all discussions are resolved.

File: collections/src/main/kotlin/ktx/collections/maps.kt
Comment (Quillraven) [reply collections/src/main/kotlin/ktx/collections/maps.kt]: I added test cases for it and documented the cases where a `java.lang.IllegalArgumentException` is thrown.

`IntMap` does not allow null keys and `GdxArrayMap` somehow supports null keys :D

File: collections/src/test/kotlin/ktx/collections/MapsTest.kt
Comment (czyzby) [inline collections/src/test/kotlin/ktx/collections/MapsTest.kt:430]: In Kotlin we can use `shouldThrow<IllegalArgumentException>` to assert if a certain exception is thrown. In this case it is fine as is, but in general `shouldThrow` is a bit more flexible, as you don't have to end the test with an exception and you can continue with other asserts.

**Comment by czyzby** [comment]
Thanks!

File: collections/src/test/kotlin/ktx/collections/MapsTest.kt
Comment (czyzby) [reply collections/src/test/kotlin/ktx/collections/MapsTest.kt:430]: See the latest commit.

Comment (Quillraven) [reply collections/src/test/kotlin/ktx/collections/MapsTest.kt:430]: Thank you for the hint. I was looking for assertThrows and something similar and then I found that approach in one of the existing test cases in ktx und just reused it.

Will keep that in mind for the future  :)

**Pull request merged**
